PINCache/PINOperationQueue.m

```diff
@@ -0,0 +1,217 @@
+//
+//  PINOperationQueue.m
+//  Pods
+//
+//  Created by Garrett Moon on 8/23/16.
+//
+//
+
+#import "PINOperationQueue.h"
+#import <pthread.h>
+
+@interface NSNumber (PINOperationQueue) <PINOperationReference>
+
+@end
+
+@interface PINOperationQueue () {
+  pthread_mutex_t _lock;
+  //increments with every operation to allow cancelation
+  NSUInteger _operationReferenceCount;
+  
+  dispatch_queue_t _serialQueue;
+  BOOL _serialQueueBusy;
+  
+  dispatch_semaphore_t _concurrentSemaphore;
+  dispatch_queue_t _concurrentQueue;
```

Comment:
Do we need to use a separate queue instead of the system global concurrent queue?

Is there any way to avoid a third queue for the semaphore?  Haven't fully reviewed the code yet (my bad), just thinking that this could be a really fantastic primitive to make wider use of in cases we may create a lot of them (like layout systems) and curious if any of these can be safely unified.